### DISCUSSION TITLE
组件API设计规范讨论（欢迎参与讨论）

### DISCUSSION THREAD

**Discussion by GaoNeng-wWw** [DevCloudFE/vue-devui#74] [General]
## 为什么创建
  - 现在共有66各组件，各组件的API命名可谓是百花齐放，百家争鸣。
  - 为了给开发者更加便利的开发体验，我们应该适当的统一命名规范

**Comment by kagol** [comment]
考虑从以下方面制定组件的api规范：
1. 使用子组件(和插槽)代替部分api：能拆分子组件的使用子组件，少使用api或json格式，子组件类似原生的html写法，能尽可能地保证组件的灵活性
2. 命名方式用中划线风格：组件的props统一使用中划线分隔
3. 组件名称前缀：组件的命名统一使用`D`前缀，指令不使用前缀和后缀
4. 样式前缀：组件的样式统一使用`devui-`前缀
5. 组件事件：区分组件的事件（写在emit里面的）和钩子（写在props里面的，钩子一般会是`before-xxx`或者`after-xxx`）
6. ...（欢迎补充）

**Comment by xingyan95** [comment]
1. 不同组件的相同功能api命名是否应该保持一致？比如【是否显示动画】参数，有的是用`animate`参数，有的是用`showAnimation`参数。其他比如设置大小、设置动画持续时长、设置自动消失时长等参数同理。
2. 在与时长相关的参数命名上，针对`duration`、`interval`、`delay`的区别，个人的一些理解，可以讨论。`duration`表示持续时间，持续一段时间后变为另一种状态，例如动画的持续时间；`interval`表示间隔，例如幻灯片自动切换的间隔时间；`delay`表示延迟，例如鼠标从某个元素移入移出，需要延迟一段时间显示隐藏。
3. 针对类似`is-xxx`和`show-xxx`的`boolean`类型参数，是不是可以把前缀去掉？直接用`xxx`表示。
4. 针对需要双向绑定的参数，直接用`v-model`是不是更好一些？避免增加额外参数，例如`v-model:xxx`。
5. `autoFocus`参数命名统一改为`autofocus`。
6. 禁用类型的参数命名统一用`disabled`。
7. 尽量避免使用通用性较低的单词缩写。

**Reply by kagol** [reply]
赞同，这块需要统一下

**Comment by kagol** [comment]
一部分规范通过贡献指南中的开发规范承载：
[https://vue-devui.github.io/contributing/](https://vue-devui.github.io/contributing/)

还有一部分不容易通过规范承载的，比如 api 的设计和命名，需要大家一起参与讨论和优化。
